src/containers/user/UserSettings.js

```diff
@@ -0,0 +1,270 @@
+  }
+
+  validate () {
+    let { errors } = this.state
+
+    if (errors.email) {
+      window.alert('Please provide a valid email.')
+      this.refs.email.focus()
+      return
+    }
+
+    if (errors.password) {
+      window.alert('Please provide a valid password.')
+      this.refs.password.focus()
+      return
+    }
+
+    return true
+  }
+
+  setEmail = event => {
+    return this.setState({
+      edited: {...this.state.edited, email: event.target.value},
+      errors: {...this.state.errors, email: !validator.isEmail(event.target.value)}
+    })
+  }
+
+  setPassword = event => {
+    return this.setState({
+      edited: {...this.state.edited, password: event.target.value},
+      errors: {...this.state.errors, password: !event.target.value}
+    })
+  }
+
+  save = (field) => {
+    if (!this.validate()) return
+
+    let { dispatch, currentUser } = this.props
+    let { editing, edited } = this.state
+    this.setState({editing: {...editing, [field]: false}})
+    dispatch(updateUserSettings({...currentUser, ...edited}, {[field]: currentUser[field]}))
+  }
+
+  edit (field) {
+    let { currentUser } = this.props
+    let { editing, edited } = this.state
+    edited[field] = currentUser[field]
+    this.setState({editing: {...editing, [field]: true}})
+  }
+
+  cancelEdit (field) {
+    let { editing } = this.state
+    this.setState({editing: {...editing, [field]: false}})
+  }
+
+  update (field, value) {
+    let { dispatch, currentUser } = this.props
+    var updatedUser = {...currentUser, [field]: value}
+    dispatch(updateUserSettings(updatedUser, {[field]: currentUser[field]}))
+  }
+
+  updateSetting (setting, value) {
+    let { currentUser } = this.props
+    let updatedSettings = {...currentUser.settings, [setting]: value}
+    this.update('settings', updatedSettings)
+  }
+
+  toggle (field) {
+    this.update(field, !this.props.currentUser[field])
+  }
+
+  toggleSetting (setting) {
+    this.updateSetting(setting, !this.props.currentUser.settings[setting])
+  }
+
+  joinCommunity () {
+    var inviteCode = window.prompt('Enter the code that was given to you by your community manager.')
+    console.log(inviteCode)
+  }
+
+  leaveCommunity (communityId) {
+    let { dispatch } = this.props
+    if (!window.confirm('Are you sure you want to leave this community?')) return
+    dispatch(leaveCommunity(communityId))
+  }
+
+  componentDidMount () {
+    let { location: { query } } = this.props
+    let { expand } = query || {}
+    switch (expand) {
+      case 'password':
+        this.setState({expand1: true, editing: {password: true}})
+        break
+      case 'prompts':
+        this.setState({expand1: true})
+        break
+      default:
+        this.setState({expand2: true})
+        break
+    }
+  }
+
+  render () {
+    let { currentUser } = this.props
+    let { editing, edited, errors, expand1, expand2, expand3 } = this.state
+
+    return <div id='user'>
+      <div className='settings'>
```

Comment:
oh man -- i really would rather not carry over all this old HTML... oh well, i guess we can clean this up later